common/src/main/java/io/netty/util/concurrent/AbstractEventExecutor.java

```diff
@@ -165,4 +166,23 @@ protected static void safeExecute(Runnable task) {
             logger.warn("A task raised an exception. Task: {}", task, t);
         }
     }
+
+    /**
+     * Like {@link #execute(Runnable)} but does not guarantee the task will be run until either
+     * a non-lazy task is executed or the executor is shut down.
+     *
+     * This is equivalent to submitting a {@link EventExecutor.LazyRunnable} to
+     * {@link #execute(Runnable)} but for an arbitrary {@link Runnable}.
+     */
+    @UnstableApi
+    public void lazyExecute(Runnable task) {
+        execute(task);
```

Comment:
shouldn't it be wrapped in LazyRunnable..?

Reply:
I don't think that should be necessary - since `LazyRunnable` is a new interface, if some subclass was to support it then it could/would also support `lazyExecute` and there would be no need for the extra wrapping (the motivation for having a separate method in the first place).

Reply:
@johnou is this resolved ?

Reply:
hard to follow the code flow in GitHub, my main point is there is a method called lazyExecute that takes Runnable, which invokes execute directly.. I would have thought we'd need something like this.. 

```java
    @UnstableApi
    public void lazyExecute(final Runnable task) {
        if (task instanceof LazyRunnable) {
            execute(task);
        } else {
            execute(new LazyRunnable() {
                @Override
                public void run() {
                    task.run();
                }
            });
        }
    }
```

Reply:
@njhill ptal

Reply:
@johnou it is not an implementation per se but rather just default behaviour for (possibly pre-existing) `AbstractEventExecutor` subclasses which don't support this lazy execution semantic. I.e. like `default` interface method in java8+. The method itself is just for providing a "hint" to allow for internal optimization (delayed wakeup) ... there should never be any problem _functionally_ with ignoring the lazy aspect and running immediately.

The idea is to provide some flexiblity in how the laziness is communicated - it can be via use of this method _and/or_ by implementing `LazyRunnable`. The former mechanism allows for deciding this for arbitrary tasks at the time of submission without having to wrap in a separate Runnable. Supporting implementations would recognize both (see the changes to `SingleThreadEventExecutor`).

Reply:
Ok so the default implementation breaks the API contract to maintain backwards compatibility.

Reply:
No API contracts are being broken :) The regular `execute` method implies "stronger" guarantees that `lazyExecute` does (i.e. that the execution will be scheduled immediately), and thus there is no problem defaulting to existing implementations of that... it's a perfectly valid way to perform `lazyExecute`. It just doesn't happen to use the extra slack afforded to it by the "lazy" part.

Reply:
I was more concerned about the javadoc, would you be against updating the wording from "Like {@link #execute(Runnable)} but does not guarantee the task will be run until either
     * a non-lazy task is executed or the executor is shut down.
     *" to something like.. "Like {@link #execute(Runnable)} but does not guarantee the task will be run until either
     * a non-lazy task is executed or the executor is shut down. If the implementing AbstractEventExecutor does not support lazy tasks it will fallback to the default execute behaviour.
     *"

Reply:
@njhill ^^?